opalog/opalog.peg

```diff
@@ -5,19 +5,208 @@ package opalog
 
+// parseConstantRule attempts to return a rule from a Body.
+// Equality expressions of the form <var> = <ground term> can be
+// converted into rules of the form <var> = <ground term> :- true.
+// This is a concise way of defining constants inside modules.
+// This function handles the conversion.
+func parseConstantRule(stmt Body) (*Rule, error) {
+    if len(stmt) > 1 {
+        return nil, fmt.Errorf("expression must be contained inside rule: %v", stmt)
+    } else if len(stmt) == 1 {
+        stmt := stmt[0]
+        if !stmt.IsEquality() {
+            return nil, fmt.Errorf("non-equality expression must be contained inside rule: %v", stmt)
+        }
+        if !stmt.Terms[2].IsGround() {
+            return nil, fmt.Errorf("constant rule value must be ground: %v", stmt)
+        }
+        switch name := stmt.Terms[1].Value.(type) {
+        case Var:
+            rule := &Rule{
+                Location: stmt.Location,
+                Name: name,
+                Value: stmt.Terms[2],
+                Body: []*Expr{
+                    &Expr{Terms: []*Term{BooleanTerm(true)}},
+                },
+            }
+            return rule, nil
+        default:
+            return nil, fmt.Errorf("rule name must be a variable: %v", stmt)
+        }
+    } else {
+        panic("unreachable")
+    }
+}
+
+}
+
+Module <- _ head:Stmt tail:( ws Stmt )* _ EOF {
+
+    // Check if multiple statements were given.
+    // If a single statement is provided, we return it as is without
+    // creating a Module object. This supports callers that only
+    // require a single statement (e.g., REPL, tests, etc.)
+    if tail == nil {
+        return head, nil
+    }
+    tailSlice := tail.([]interface{})
+    if len(tailSlice) == 0 {
+        return head, nil
+    }
+
+    // Multiple statements were provided so create a new Module.
+    mod := &Module{
+        Package: head.(*Package),
+    }
+    for _, s := range tailSlice {
+        switch stmt := s.([]interface{})[1].(type) {
+        case *Import:
+            mod.Imports = append(mod.Imports, stmt)
+        case *Rule:
+            mod.Rules = append(mod.Rules, stmt)
+        case Body:
+            rule, err := parseConstantRule(stmt)
+            if err != nil {
+                return nil, err
+            }
+            mod.Rules = append(mod.Rules, rule)
+        }
+    }
+    return mod, nil
+}
+
+Stmt <- val:(Package / Import / Rule / Body / Comment) {
+    return val, nil
+}
+
+Package <- "package" ws val:(Ref / Var) {
+    // All packages are implicitly declared under the default root document.
+    path := RefTerm(DefaultRootDocument)
+    switch v := val.(*Term).Value.(type) {
+    case Ref:
+        // Convert head of package Ref to String because it will be prefixed
+        // with the root document variable.
+        head := v[0]
+        head = StringTerm(string(head.Value.(Var)))
+        head.Location = v[0].Location
+        tail := v[1:]
+        if !tail.IsGround() {
```

Comment:
Seems this allows us to write package a['b']['c'] as well as a.b.c.  That's the right thing to do so that we can handle the case of a key containing spaces.  Let's add a comment explaining that and why that forces us to have the IsGround check.